### PR TITLE
proper executor/block type for benchmarks and try-runtime

### PR DESCRIPTION
**Pull request by kianenigma**
closes https://github.com/paritytech/polkadot/issues/2516

What I don't get is that when I run these two commands with `--chain=polkadot` and inspect the executor logs, I get `polkadot-0` as the spec-version, but when I do `--chain=polkadot-dev` I get `polkadot-30`. Is this why we also run benchmarks always with `-dev` variants? And why is this? 

Either way, that coded needed refactoring.

### CODE REVIEW

**Comment by shawntabrizi** [comment]
I think there is a typo in your post. Do you mean `--chain=polkadot-dev`?

`--chain=polkadot` uses the hardcoded genesis with a specific wasm file which DOES NOT contain benchmarking or your try-runtime apis. So `--chain=polkadot` will not work for this kind of stuff, and is a pretty common error.

File: cli/src/command.rs [hunk 1]
```diff
@@ -268,13 +268,27 @@ pub fn run() -> Result<()> {
...
+				Ok(runner.sync_run(|config| {
+					cmd.run::<service::polkadot_runtime::Block, service::PolkadotExecutor>(config)
+						.map_err(|e| Error::SubstrateCli(e))
+				})?)
+			} else if spec_name.contains("kusama") {
+				Ok(runner.sync_run(|config| {
+					cmd.run::<service::kusama_runtime::Block, service::KusamaExecutor>(config)
+						.map_err(|e| Error::SubstrateCli(e))
+				})?)
+			} else if spec_name.contains("westend") {
```
Comment (shawntabrizi) [inline cli/src/command.rs:284]: maybe we can fix that common problem by only allowing benchmarks with `*-dev`

**Review by shawntabrizi — APPROVED** [review]
👍

**Comment by kianenigma** [comment]
> --chain=polkadot uses the hardcoded genesis with a specific wasm file which DOES NOT contain benchmarking or your try-runtime apis

Which hardcoded genesis exactly? 

I see, so my observation was expected, but I still don't get why we cannot compile the real polkadot node+runtime, with these additional apis piggybacked?

File: cli/src/command.rs (hunk 1 above)
Comment (kianenigma) [reply cli/src/command.rs:284]: Given your explanation, I'll probabkly do that. So we never ever ever want to run benchmarks with any chain other than `*-dev`?

Comment (shawntabrizi) [reply cli/src/command.rs:284]: probably not.

**Comment by kianenigma** [comment]
> > Which hardcoded genesis exactly?
> 
> The genesis of the real Polkadot blockchain :)
> 
> https://github.com/paritytech/polkadot/tree/master/node/service/res

Huh, I did _not_ know that existed :D

File: cli/src/command.rs [hunk 2]
```diff
@@ -268,13 +268,27 @@ pub fn run() -> Result<()> {
...
-
 			set_default_ss58_version(chain_spec);
-
-			Ok(runner.sync_run(|config| {
-				cmd.run::<service::kusama_runtime::Block, service::KusamaExecutor>(config)
-				.map_err(|e| Error::SubstrateCli(e))
-			})?)
+			let spec_name = runner.config().impl_name.to_lowercase();
+
+			if spec_name.contains("polkadot") {
```
Comment (bkchr) [inline cli/src/command.rs:274]:
```suggestion
			if spec.is_polkadot() {
```

Please use this everywhere. There is the trait `IdentifyVariant` in `polkadot-service` that you will need to import for this.

**Review by bkchr — CHANGES_REQUESTED** [review]

File: cli/src/command.rs (hunk 2 above)
Comment (kianenigma) [reply cli/src/command.rs:274]: I would rather tweak the trait to be able to assert `spec.is_polkadot_dev()`, if you don't mind.

Comment (shawntabrizi) [reply cli/src/command.rs:274]: `is_polkadot(Option<suffix>)`

Comment (kianenigma) [reply cli/src/command.rs:274]: I was thinking of adding a `is_dev()` that return true on anything ending with `*-dev`.

**Review by bkchr — APPROVED** [review]

**Pull request merged**